api-authentication/server/server.ts

```diff
@@ -0,0 +1,194 @@
+/* eslint-disable @typescript-eslint/no-unused-vars -- Remove me */
+import 'dotenv/config';
+import pg from 'pg';
+import argon2 from 'argon2';
+import express from 'express';
+import jwt from 'jsonwebtoken';
+import { ClientError, errorMiddleware } from './lib/index.js';
+
+type User = {
+  userId: number;
+  username: string;
+  hashedPassword: string;
+};
+type Entry = {
+  entryId: number;
+  title: string;
+  notes: string;
+  photoUrl: string;
+};
+type Auth = {
+  username: string;
+  password: string;
+};
+
+const db = new pg.Pool({
+  connectionString: process.env.DATABASE_URL,
+  ssl: {
+    rejectUnauthorized: false,
+  },
+});
+
+const hashKey = process.env.TOKEN_SECRET;
+if (!hashKey) throw new Error('TOKEN_SECRET not found in .env');
+
+const app = express();
+app.use(express.json());
+
+app.post('/api/auth/sign-up', async (req, res, next) => {
+  try {
+    const { username, password } = req.body as Partial<Auth>;
+    if (!username || !password) {
+      throw new ClientError(400, 'username and password are required fields');
+    }
+    const hashedPassword = await argon2.hash(password);
+    const sql = `
+      insert into "users" ("username", "hashedPassword")
+      values ($1, $2)
+      returning *
+    `;
+    const params = [username, hashedPassword];
+    const result = await db.query<User>(sql, params);
+    const [user] = result.rows;
+    res.status(201).json(user);
+  } catch (err) {
+    next(err);
+  }
+});
+
+app.post('/api/auth/sign-in', async (req, res, next) => {
+  try {
+    const { username, password } = req.body as Partial<Auth>;
+    if (!username || !password) {
+      throw new ClientError(401, 'invalid login');
+    }
+
+    /* your code starts here */
+    const sql = `
+          select "userId",
+                 "hashedPassword"
+           from  "users"
+           where "username" = $1;
+           `;
+    const params = [username];
+    const result = await db.query(sql, params);
+    if (!result) throw new ClientError(401, 'invalid login');
+    const { userId, hashedPassword } = result.rows[0];
+    const isMatching = await argon2.verify(hashedPassword, password);
+    if (!isMatching) throw new ClientError(401, 'invalid login');
+    const payload = {
+      userId,
+      username,
+    };
+    const token = jwt.sign(payload, 'snitchesgetstiches');
```

Comment:
You need to use the value in the `.env` file. That way your secret never gets submitted into GitHub, where it can be easily stolen and used to hack your site.